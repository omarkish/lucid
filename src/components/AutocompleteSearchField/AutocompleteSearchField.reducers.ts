import * as DropMenu from '../DropMenu/DropMenu.reducers';
import SearchField from '../SearchField/SearchField.reducers';
import { IAutocompleteSearchFieldState } from './AutocompleteSearchField';

export function onSelect(
	state: IAutocompleteSearchFieldState,
	selectedIndex: number | null
) {
	return {
		...state,
		selectedIndex,
	};
}

export function onSearch(
	state: IAutocompleteSearchFieldState,
	searchText: string,
	firstVisibleIndex: number
) {
	return {
		...state,
		searchText,
		DropMenu: {
			...DropMenu.onFocusOption(state.DropMenu, firstVisibleIndex),
		},
	};
}

export { DropMenu, SearchField };
