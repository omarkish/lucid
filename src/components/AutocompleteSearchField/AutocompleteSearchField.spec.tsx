import * as _ from 'lodash';
import React from 'react';
import { shallow } from 'enzyme';
import { common } from '../../util/generic-tests';
import { AutocompleteSearchFieldDumb as AutocompleteSearchField } from './AutocompleteSearchField.reducers';
import { DropMenuDumb as DropMenu } from '../DropMenu/DropMenu';

const { Option } = AutocompleteSearchField;

describe('AutocompleteSearchField', () => {
	common(AutocompleteSearchField, {
		exemptFunctionProps: ['optionFilter'] as any,
	});

	describe('render', () => {
		it('should render selections', () => {
			const wrapper = shallow(
				<AutocompleteSearchField>
					<Option>option a</Option>
					<Option>option b</Option>
					<Option>option c</Option>
				</AutocompleteSearchField>
			);

			expect(wrapper.find(DropMenu).length).toBe(1);
		});

		it('should pass `isDisabled` to `Options`', () => {
			const wrapper = shallow(
				<AutocompleteSearchField>
					<Option isDisabled>option a</Option>
					<Option>option b</Option>
					<Option>option c</Option>
				</AutocompleteSearchField>
			);
			const [first, second, third] = wrapper
				.find(DropMenu.Option)
				.map((option) => option.prop('isDisabled'));
			expect(first).toBe(true);
			expect(second).toBe(false);
			expect(third).toBe(false);
		});
	});

	describe('props', () => {
		describe('onSearch', () => {
			it('should work', () => {
				const onSearch = jest.fn();
				const wrapper: any = shallow(
					<AutocompleteSearchField onSearch={onSearch}>
						<Option callbackId={'zero'}>Zero</Option>
						<Option callbackId={'one'}>One</Option>
					</AutocompleteSearchField>
				);
				const expected = {
					event: 'fake',
					props: {
						callbackId: 'zero',
						children: 'Zero',
						isDisabled: false,
						isHidden: false,
						isWrapped: true,
					},
				};

				wrapper.find('SearchField').prop('onChange')('ero', { event: 'fake' });

				expect(onSearch).toHaveBeenCalledWith('ero', 0, expected);
			});
		});

		describe('onSelect', () => {
			it('should work when fired from the DropMenu', () => {
				const onSelect = jest.fn();
				const mockSelectionCallback = {
					event: {
						preventDefault: _.noop,
					},
					props: {},
				};

				const wrapper: any = shallow(
					<AutocompleteSearchField onSelect={onSelect}>
						<Option>option a</Option>
					</AutocompleteSearchField>
				);

				wrapper.find('DropMenu').first().prop('onSelect')(
					10,
					mockSelectionCallback
				);

				expect(onSelect).toHaveBeenCalledWith(10, mockSelectionCallback);
			});

			it('should work when selecting filtered options', () => {
				const onSelect = jest.fn();
				const mockSelectionCallback = {
					event: { stopPropagation: _.noop, preventDefault: _.noop },
					props: {},
				};

				const wrapper: any = shallow(
					<AutocompleteSearchField onSelect={onSelect}>
						<Option isHidden>option a</Option>
						<Option>option b</Option>
					</AutocompleteSearchField>
				);

				wrapper.find('DropMenu').first().prop('onSelect')(
					0,
					mockSelectionCallback
				);

				expect(onSelect).toHaveBeenCalledWith(0, mockSelectionCallback);
			});

			it('should work when fired from the selection removal ', () => {
				const onSelect = jest.fn();
				const mockSelectionCallback = {
					event: {
						preventDefault: _.noop,
					},
					props: {},
				};

				const wrapper: any = shallow(
					<AutocompleteSearchField
						onSelect={onSelect}
						onSearch={_.noop}
						selectedIndex={0}
					>
						<Option>option a</Option>
					</AutocompleteSearchField>
				);

				wrapper.find('Selection').first().prop('onRemove')(
					mockSelectionCallback
				);

				expect(onSelect).toHaveBeenCalledWith(null, mockSelectionCallback);
			});
		});

		describe('Error', () => {
			it('should apply the appropriate classNames to the saerch', () => {
				const wrapper = shallow(
					<AutocompleteSearchField Error={'Erroring out'}>
						<Option>option a</Option>
						<Option>option b</Option>
					</AutocompleteSearchField>
				);

				const searchWrapper = wrapper.find(
					'.lucid-AutocompleteSearchField-search-is-error'
				);

				expect(searchWrapper.exists()).toBeTruthy();
			});

			it('should render out the error div', () => {
				const wrapper = shallow(
					<AutocompleteSearchField Error={'Erroring out'}>
						<Option>option a</Option>
						<Option>option b</Option>
					</AutocompleteSearchField>
				);

				const searchWrapper = wrapper.find(
					'.lucid-AutocompleteSearchField-error-content'
				);

				expect(searchWrapper.text()).toEqual('Erroring out');
			});

			it('should not render the error div', () => {
				const wrapper = shallow(
					<AutocompleteSearchField Error={true}>
						<Option>option a</Option>
						<Option>option b</Option>
					</AutocompleteSearchField>
				);

				const searchWrapper = wrapper.find(
					'.lucid-AutocompleteSearchField-search-is-error'
				);
				const errorWrapper = wrapper.find(
					'.lucid-AutocompleteSearchField-error-content'
				);
				expect(errorWrapper.exists()).toBeFalsy();
				expect(searchWrapper).toBeTruthy();
			});
		});
	});

	describe('custom formatting', () => {
		it('should render Option child function by passing in {searchText}, setting filterText on each option and using a custom optionFilter', () => {
			const optionFilter = (searchText: any, { filterText }: any) => {
				if (filterText) {
					return new RegExp(_.escapeRegExp(searchText), 'i').test(filterText);
				}
				return true;
			};

			expect(
				shallow(
					<AutocompleteSearchField optionFilter={optionFilter} searchText='tion'>
						<Option name='OptionA' Selected='option a' filterText='option a'>
							{({ searchText }: any) => (
								<div style={{ display: 'flex' }}>
									<div style={{ width: 100 }}>{searchText}</div>
									<div>option a</div>
								</div>
							)}
						</Option>
						<Option name='OptionB' Selected='option b' filterText='option b'>
							{({ searchText }: any) => (
								<div style={{ display: 'flex' }}>
									<div style={{ width: 100 }}>{searchText}</div>
									<div>option b</div>
								</div>
							)}
						</Option>
						<Option name='OptionC' Selected='option c' filterText='option c'>
							{({ searchText }: any) => (
								<div style={{ display: 'flex' }}>
									<div style={{ width: 100 }}>{searchText}</div>
									<div>option c</div>
								</div>
							)}
						</Option>
					</AutocompleteSearchField>
				)
			).toMatchSnapshot();
		});
	});
});
