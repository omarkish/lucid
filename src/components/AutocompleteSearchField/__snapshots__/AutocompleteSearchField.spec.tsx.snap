// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteSearchField [common] example testing should match snapshot(s) for Asynchronous 1`] = `
<section>
  <div
    class="lucid-AutocompleteSearchField"
  >
    <div
      class="lucid-DropMenu lucid-DropMenu-base lucid-AutocompleteSearchField-DropMenu lucid-AutocompleteSearchField-DropMenu-is-small"
    >
      <span
        class="lucid-ContextMenu"
      >
        <div
          class="lucid-DropMenu-Control"
          tabindex="0"
        >
          <div
            class="lucid-SearchField lucid-AutocompleteSearchField-search"
          >
            <input
              autocomplete="off"
              class="lucid-TextField lucid-TextField-is-single-line"
              placeholder="Type here to simulate an API backed search"
              rows="5"
              type="text"
              value=""
            />
            <div
              class="lucid-SearchField-Icon-container"
            >
              <svg
                class="lucid-Icon lucid-Icon-color-primary lucid-SearchIcon lucid-SearchField-Icon"
                height="12"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="12"
              >
                <circle
                  cx="6"
                  cy="6"
                  r="5.5"
                />
                <path
                  d="M15.5 15.5L9.876 9.876"
                />
              </svg>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</section>
`;

exports[`AutocompleteSearchField [common] example testing should match snapshot(s) for Default 1`] = `
<section>
  <div
    class="lucid-AutocompleteSearchField"
    style="max-width:200px"
  >
    <div
      class="lucid-DropMenu lucid-DropMenu-base lucid-AutocompleteSearchField-DropMenu lucid-AutocompleteSearchField-DropMenu-is-small"
    >
      <span
        class="lucid-ContextMenu"
      >
        <div
          class="lucid-DropMenu-Control"
          tabindex="0"
        >
          <div
            class="lucid-SearchField lucid-AutocompleteSearchField-search"
          >
            <input
              autocomplete="off"
              class="lucid-TextField lucid-TextField-is-single-line"
              placeholder="Search State"
              rows="5"
              type="text"
              value=""
            />
            <div
              class="lucid-SearchField-Icon-container"
            >
              <svg
                class="lucid-Icon lucid-Icon-color-primary lucid-SearchIcon lucid-SearchField-Icon"
                height="12"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="12"
              >
                <circle
                  cx="6"
                  cy="6"
                  r="5.5"
                />
                <path
                  d="M15.5 15.5L9.876 9.876"
                />
              </svg>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</section>
`;

exports[`AutocompleteSearchField [common] example testing should match snapshot(s) for FormattedOptions 1`] = `
<section>
  <div
    class="lucid-AutocompleteSearchField"
    style="max-width:200px"
  >
    <div
      class="lucid-DropMenu lucid-DropMenu-base lucid-AutocompleteSearchField-DropMenu lucid-AutocompleteSearchField-DropMenu-is-small"
    >
      <span
        class="lucid-ContextMenu"
      >
        <div
          class="lucid-DropMenu-Control"
          tabindex="0"
        >
          <div
            class="lucid-SearchField lucid-AutocompleteSearchField-search"
          >
            <input
              autocomplete="off"
              class="lucid-TextField lucid-TextField-is-single-line"
              placeholder="Search Options"
              rows="5"
              type="text"
              value=""
            />
            <div
              class="lucid-SearchField-Icon-container"
            >
              <svg
                class="lucid-Icon lucid-Icon-color-primary lucid-SearchIcon lucid-SearchField-Icon"
                height="12"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="12"
              >
                <circle
                  cx="6"
                  cy="6"
                  r="5.5"
                />
                <path
                  d="M15.5 15.5L9.876 9.876"
                />
              </svg>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</section>
`;

exports[`AutocompleteSearchField [common] example testing should match snapshot(s) for GroupedOptions 1`] = `
<section>
  <div
    class="lucid-AutocompleteSearchField"
    style="max-width:200px"
  >
    <div
      class="lucid-DropMenu lucid-DropMenu-base lucid-AutocompleteSearchField-DropMenu lucid-AutocompleteSearchField-DropMenu-is-small"
    >
      <span
        class="lucid-ContextMenu"
      >
        <div
          class="lucid-DropMenu-Control"
          tabindex="0"
        >
          <div
            class="lucid-SearchField lucid-AutocompleteSearchField-search"
          >
            <input
              autocomplete="off"
              class="lucid-TextField lucid-TextField-is-single-line"
              rows="5"
              type="text"
              value=""
            />
            <div
              class="lucid-SearchField-Icon-container"
            >
              <svg
                class="lucid-Icon lucid-Icon-color-primary lucid-SearchIcon lucid-SearchField-Icon"
                height="12"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="12"
              >
                <circle
                  cx="6"
                  cy="6"
                  r="5.5"
                />
                <path
                  d="M15.5 15.5L9.876 9.876"
                />
              </svg>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</section>
`;

exports[`AutocompleteSearchField [common] example testing should match snapshot(s) for InvalidOption 1`] = `
<div
  class="lucid-AutocompleteSearchField"
>
  <div
    class="lucid-DropMenu lucid-DropMenu-base lucid-AutocompleteSearchField-DropMenu lucid-AutocompleteSearchField-DropMenu-is-small"
  >
    <span
      class="lucid-ContextMenu"
    >
      <div
        class="lucid-DropMenu-Control"
        tabindex="0"
      >
        <div
          class="lucid-SearchField lucid-AutocompleteSearchField-search lucid-AutocompleteSearchField-search-is-error"
        >
          <input
            autocomplete="off"
            class="lucid-TextField lucid-TextField-is-single-line"
            rows="5"
            type="text"
            value=""
          />
          <div
            class="lucid-SearchField-Icon-container"
          >
            <svg
              class="lucid-Icon lucid-Icon-color-primary lucid-SearchIcon lucid-SearchField-Icon"
              height="12"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width="12"
            >
              <circle
                cx="6"
                cy="6"
                r="5.5"
              />
              <path
                d="M15.5 15.5L9.876 9.876"
              />
            </svg>
          </div>
        </div>
      </div>
    </span>
  </div>
  <div
    class="lucid-AutocompleteSearchField-error-content"
  >
    Please Choose Green
  </div>
</div>
`;

exports[`AutocompleteSearchField [common] example testing should match snapshot(s) for Props 1`] = `
<section>
  <h5>
    Loading
  </h5>
  <div
    class="lucid-AutocompleteSearchField"
  >
    <div
      class="lucid-DropMenu lucid-DropMenu-base lucid-AutocompleteSearchField-DropMenu lucid-AutocompleteSearchField-DropMenu-is-small"
    >
      <span
        class="lucid-ContextMenu"
      >
        <div
          class="lucid-DropMenu-Control"
          tabindex="0"
        >
          <div
            class="lucid-SearchField lucid-AutocompleteSearchField-search"
          >
            <input
              autocomplete="off"
              class="lucid-TextField lucid-TextField-is-single-line"
              placeholder=""
              rows="5"
              type="text"
              value=""
            />
            <div
              class="lucid-SearchField-Icon-container"
            >
              <svg
                class="lucid-Icon lucid-Icon-color-primary lucid-SearchIcon lucid-SearchField-Icon"
                height="12"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="12"
              >
                <circle
                  cx="6"
                  cy="6"
                  r="5.5"
                />
                <path
                  d="M15.5 15.5L9.876 9.876"
                />
              </svg>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
  <h5>
    Disabled
  </h5>
  <div
    class="lucid-AutocompleteSearchField"
  >
    <div
      class="lucid-DropMenu lucid-DropMenu-base lucid-AutocompleteSearchField-DropMenu lucid-AutocompleteSearchField-DropMenu-is-small"
    >
      <span
        class="lucid-ContextMenu"
      >
        <div
          class="lucid-DropMenu-Control"
        >
          <div
            class="lucid-SearchField lucid-AutocompleteSearchField-search"
          >
            <input
              autocomplete="off"
              class="lucid-TextField lucid-TextField-is-disabled lucid-TextField-is-single-line"
              disabled=""
              placeholder=""
              rows="5"
              type="text"
              value=""
            />
            <div
              class="lucid-SearchField-Icon-container lucid-SearchField-Icon-is-disabled"
            >
              <svg
                class="lucid-Icon lucid-Icon-color-primary lucid-SearchIcon lucid-SearchField-Icon"
                height="12"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="12"
              >
                <circle
                  cx="6"
                  cy="6"
                  r="5.5"
                />
                <path
                  d="M15.5 15.5L9.876 9.876"
                />
              </svg>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</section>
`;

exports[`AutocompleteSearchField [common] example testing should match snapshot(s) for SelectedOption 1`] = `
<section>
  <div
    class="lucid-AutocompleteSearchField"
    style="max-width:200px"
  >
    <div
      class="lucid-DropMenu lucid-DropMenu-base lucid-AutocompleteSearchField-DropMenu lucid-AutocompleteSearchField-DropMenu-is-small"
    >
      <span
        class="lucid-ContextMenu"
      >
        <div
          class="lucid-DropMenu-Control"
          tabindex="0"
        >
          <div
            class="lucid-SearchField lucid-AutocompleteSearchField-search"
          >
            <input
              autocomplete="off"
              class="lucid-TextField lucid-TextField-is-single-line"
              placeholder="Search Color"
              rows="5"
              type="text"
              value=""
            />
            <div
              class="lucid-SearchField-Icon-container"
            >
              <svg
                class="lucid-Icon lucid-Icon-color-primary lucid-SearchIcon lucid-SearchField-Icon"
                height="12"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="12"
              >
                <circle
                  cx="6"
                  cy="6"
                  r="5.5"
                />
                <path
                  d="M15.5 15.5L9.876 9.876"
                />
              </svg>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</section>
`;

exports[`AutocompleteSearchField custom formatting should render Option child function by passing in {searchText}, setting filterText on each option and using a custom optionFilter 1`] = `
<div
  className="lucid-AutocompleteSearchField"
>
  <DropMenu
    ContextMenu={
      Object {
        "alignmentOffset": -13,
        "directonOffset": -1,
        "minWidthOffset": -28,
      }
    }
    alignment="start"
    className="lucid-AutocompleteSearchField-DropMenu lucid-AutocompleteSearchField-DropMenu-is-small"
    direction="down"
    flyOutStyle={
      Object {
        "maxHeight": "18em",
      }
    }
    focusedIndex={null}
    isDisabled={false}
    isExpanded={false}
    onCollapse={[Function]}
    onExpand={[Function]}
    onFocusNext={[Function]}
    onFocusOption={[Function]}
    onFocusPrev={[Function]}
    onSelect={[Function]}
    optionContainerStyle={Object {}}
    portalId=""
    selectedIndices={null}
  >
    <DropMenu.Control>
      <SearchField
        autoComplete="off"
        className="lucid-AutocompleteSearchField-search"
        debounceLevel={500}
        isDisabled={false}
        onChange={[Function]}
        onChangeDebounced={[Function]}
        onSubmit={[Function]}
        value="tion"
      />
    </DropMenu.Control>
    <DropMenu.Option
      isDisabled={false}
      isHidden={false}
      isWrapped={true}
      key="AutocompleteSearchFieldOption0"
      name="OptionA"
    >
      <Component
        searchText="tion"
      />
    </DropMenu.Option>
    <DropMenu.Option
      isDisabled={false}
      isHidden={false}
      isWrapped={true}
      key="AutocompleteSearchFieldOption1"
      name="OptionB"
    >
      <Component
        searchText="tion"
      />
    </DropMenu.Option>
    <DropMenu.Option
      isDisabled={false}
      isHidden={false}
      isWrapped={true}
      key="AutocompleteSearchFieldOption2"
      name="OptionC"
    >
      <Component
        searchText="tion"
      />
    </DropMenu.Option>
  </DropMenu>
</div>
`;
